### PR TITLE
PIL-934: CYA page for BTN

### PR DESCRIPTION
**Pull request by MoDweik**
**Changes**

- Implement BTN CYA page (controller, view and specs)
- Link up CYA with the rest of the BTN journey
- Refactor and increase minimum statement coverage to 89%

### CODE REVIEW

File: app/controllers/btn/CheckYourAnswersController.scala [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+          Ok(view(summaryList))
+
+        case _ => Redirect(controllers.routes.IndexController.onPageLoad)
+
+      }).getOrElse(Redirect(controllers.routes.JourneyRecoveryController.onPageLoad()))
+    }
+  }
+
+  def onSubmit: Action[AnyContent] = (identify andThen getData andThen requireData) {
+    Redirect(controllers.routes.UnderConstructionController.onPageLoad)
```
Comment (MoDweik) [inline app/controllers/btn/CheckYourAnswersController.scala:73]: No AC around submit; Wendy advised subsequent Mongo/downstream actions will be handled on the next page (PIL-937).

File: app/controllers/btn/BtnAccountingPeriodController.scala [hunk 2]
```diff
@@ -48,36 +53,39 @@ class BtnAccountingPeriodController @Inject() (
...
-          Ok(view(list, mode, changeAccountingPeriodUrl))
-        case _ => Redirect(controllers.routes.JourneyRecoveryController.onPageLoad(None))
+        case Right(Fulfilled) if !accountStatus => Ok(viewReturnSubmitted(accountingPeriods))
+        case Right(Open) if !accountStatus      => Ok(view(accountingPeriods, mode, changeAccountingPeriodUrl))
+        case _                                  => Redirect(controllers.routes.JourneyRecoveryController.onPageLoad(None))
       }
   }
 
-  def onSubmit(mode: Mode): Action[AnyContent] = (identify andThen getData).async { implicit request =>
+  def onSubmit(mode: Mode): Action[AnyContent] = (identify andThen getSubscriptionData).async { implicit request =>
```
Comment (MoDweik) [inline app/controllers/btn/BtnAccountingPeriodController.scala:88]: The behaviour of this controller will need to be reviewed. The user is unable to change the dates of the AccountingPeriod fetched from subscription.

File: test/views/btn/CheckYourAnswersViewSpec.scala [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+class CheckYourAnswersViewSpec extends ViewSpecBase {
+
+  val page: CheckYourAnswersView = inject[CheckYourAnswersView]
+  val view: Document             = Jsoup.parse(page(summaryList)(request, appConfig, realMessagesApi.preferred(request)).toString())
+
+  "CheckYourAnswersView" must {
+
+    "have a title" in {
+      view.getElementsByTag("title").text mustEqual "Check Your Answers - Report Pillar 2 top-up taxes - GOV.UK GOV.UK"
```
Comment (MoDweik) [inline test/views/btn/CheckYourAnswersViewSpec.scala:32]: Can't remember if we're testing static or dynamic values from the Mike Pratt discussion, but static seems better at catching errors. Thoughts welcome.

File: app/controllers/CheckYourAnswersController.scala [hunk 4]
```diff
@@ -1,48 +0,0 @@
...
-import config.FrontendAppConfig
-import controllers.actions.{DataRequiredAction, DataRetrievalAction, IdentifierAction}
-import play.api.Logging
-import play.api.i18n.{I18nSupport, MessagesApi}
-import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
-import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
-import viewmodels.govuk.summarylist._
-import views.html.CheckYourAnswersView
-
-class CheckYourAnswersController @Inject() (
```
[the remark this answers is not in the document]
Comment (MoDweik) [reply app/controllers/CheckYourAnswersController.scala:29]: No, a placeholder. I moved it to the _btn_ package but Git decided to show a delete/create change.

File: app/controllers/btn/BtnAccountingPeriodController.scala [hunk 5]
```diff
@@ -48,36 +53,39 @@ class BtnAccountingPeriodController @Inject() (
     extends FrontendBaseController
     with I18nSupport {
 
-  def onPageLoad(mode: Mode): Action[AnyContent] = (identify andThen getData andThen requireData).async { implicit request =>
+  def onPageLoad(mode: Mode): Action[AnyContent] = (identify andThen getSubscriptionData andThen requireSubscriptionData).async { implicit request =>
     val changeAccountingPeriodUrl = appConfig.changeAccountingPeriodUrl
     val subAccountingPeriod       = request.subscriptionLocalData.subAccountingPeriod
-    val accountStatus             = request.subscriptionLocalData.accountStatus.map(_.inactive).getOrElse(true)
+    setSubscriptionAccountingPeriod(subAccountingPeriod).apply(request).map(_ => ())
```
[the remark this answers is not in the document]
Comment (MoDweik) [reply app/controllers/btn/BtnAccountingPeriodController.scala:59]: Agreed, but I'd rather change this in a separate ticket and have it tested properly as we introduce this significant change to the controller's behaviour, so unless you insist, I will raise a ticket for that purpose.

File: test/views/btn/CheckYourAnswersViewSpec.scala (hunk 3 above)
Comment (MoDweik) [reply test/views/btn/CheckYourAnswersViewSpec.scala:32]: There are two titles on the page, this was concatenating them. I made the change to specify the child element and changed the static response accordingly.

File: app/navigation/BtnNavigator.scala [hunk 6]
```diff
@@ -25,59 +27,44 @@ import javax.inject.{Inject, Singleton}
 @Singleton
 class BtnNavigator @Inject() {
 
-  def nextPage(page: Page, mode: Mode, userAnswers: UserAnswers): Call = mode match {
-    case NormalMode =>
-      normalRoutes(page)(userAnswers)
-    case CheckMode =>
-      checkRouteMap(page)(userAnswers)
+  def nextPage(page: Page, mode: Mode, userAnswers: UserAnswers): Call = page match {
+    case booleanPage: QuestionPage[Boolean]
```
Comment (TevynAllen) [inline app/navigation/BtnNavigator.scala:31]: Why did you get rid of NormalMode and CheckMode? You'll need this for navigation between the check your answers and pages and normal routing.

File: app/navigation/BtnNavigator.scala [hunk 7]
```diff
@@ -25,59 +27,44 @@ import javax.inject.{Inject, Singleton}
...
-      .map { provided =>
-        if (provided) {
-          controllers.btn.routes.BtnSubmitUKTRController.onPageLoad
-        } else {
-          controllers.btn.routes.CheckYourAnswersController.onPageLoad
-        }
-      }
-      .getOrElse(controllers.routes.JourneyRecoveryController.onPageLoad())
-
-  private val checkRouteMap: Page => UserAnswers => Call = {
```
Comment (TevynAllen) [inline app/navigation/BtnNavigator.scala:78]: This is the only function you needed to add too?

File: app/navigation/BtnNavigator.scala (hunk 6 above)
Comment (MoDweik) [reply app/navigation/BtnNavigator.scala:31]: The behaviour is still maintained but we now achieve more with less code. The mode is passed down to the next function (yesNoNavigator in this case), and eventually to the endpoint call.

File: app/navigation/BtnNavigator.scala (hunk 7 above)
Comment (MoDweik) [reply app/navigation/BtnNavigator.scala:78]: Addressed above, let me know what you think

File: app/controllers/btn/BtnAccountingPeriodController.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (MoDweik) [reply app/controllers/btn/BtnAccountingPeriodController.scala]: I spoke to Richard and he clarified: not storing/loading saved user answers is a static stub problem, this functionality works fine in QA/Prod. 

Also, as we now call the sessionRepository for UserAnswers, we don't require access to UserAnswers through the request and can use a SubscriptionDataRequest, and therefore no need to store AccountingPeriod in UserAnswers, which in fact was a mistake to begin with.

**Review by TevynAllen — DISMISSED** [review]

**Review by TevynAllen — APPROVED** [review]

File: conf/app.routes [hunk 8]
```diff
@@ -43,7 +41,6 @@ GET         /below-threshold-notification/accounting-period
 POST        /below-threshold-notification/accounting-period                       controllers.btn.BtnAccountingPeriodController.onSubmit(mode: Mode = NormalMode)
 GET         /below-threshold-notification/change-accounting-period                controllers.btn.BtnAccountingPeriodController.onPageLoad(mode: Mode = CheckMode)
 POST        /below-threshold-notification/change-accounting-period                controllers.btn.BtnAccountingPeriodController.onSubmit(mode: Mode = CheckMode)
-GET         /below-threshold-notification/return-submitted                        controllers.btn.BtnAccountingPeriodController.onPageLoadReturnSubmitted
```
Comment (srikanthhmrc) [inline conf/app.routes:46]: why we removed this

**Pull request merged**